CNI network plugin

dcbw opened the pull request:
@thockin 

Add an experimental network plugin implementation named "cni" that uses the Container Networking Interface (CNI) specification for configuring networking for pods.

https://github.com/appc/cni/blob/master/SPEC.md

Originally by @rajatchopra

Commit:
CNI network plugin

Add an experimental network plugin implementation named "cni" that
uses the Container Networking Interface (CNI) specification for
configuring networking for pods.

https://github.com/appc/cni/blob/master/SPEC.md

dcbw commented:
This PR is a rebase and fixup of https://github.com/kubernetes/kubernetes/pull/13766

thockin commented:
LGTM - squash and fix nit

dcbw commented:
@thockin will do

dcbw commented:
Since I'm a Red Hat employee, I'm covered under the Google Corporate CLA that Red Hat signed on Apr 29, 2015 15:23 PDT (and it is correctly listed at https://cla.developers.google.com/clas while I'm logged in).  So is @rajatchopra ...

thockin commented:
Are you in the googlegroup you need to be in to indicate your RedHatterness?  @smarterclayton

thockin commented:
shippable indicated that the cni pkg had a test error - I kicked it just in case, but can you confirm that 'make test' passes on this PR?

eparis commented:
The commit was authored using @rajatchopra personal gmail address, not his @redhat.com address. `git commit --amend --author="Author Name <email@address.com>"` and then forcepush should make the bot behave better.

pmorie commented:
@kubernetes/rh-networking

dcbw commented:
> shippable indicated that the cni pkg had a test error - I kicked it just in case, but can you confirm that 'make test' passes on this PR?

It seems to always fail on Go 1.3, but never on Go 1.4?  I can't get it to fail 'make check' locally with Go 1.4.2 on x86-64.  I'll try to dig into it...

dcbw commented:
> The commit was authored using @rajatchopra personal gmail address, not his @redhat.com address.

@eparis Fixed that.  How do we kick the CLA bot?

thockin commented:
I didn't un-lgtm #13766 and it got merged.  can you rebase to get this fixup?

bgrant0607 commented:
I'm going to revert #13766

bgrant0607 commented:
Reverted by #14544

thockin commented:
That's great, thanks.

dcbw commented:
@thockin Current status on this one: still chasing the periodic failure; I cannot get it to fail locally with either Go 1.3.3 or Go 1.4.2, but I'm diagnosing the issue.

dcbw commented:
I signed it!

thockin commented:
What happened?  The second commit in this PR is obviously wrong..

We're already late for cutoff...

dcbw commented:
> What happened? The second commit in this PR is obviously wrong..
> 
> We're already late for cutoff...

@thockin Shippable is still failing mostly in the Go 1.3 tests, with "write |1: broken pipe" when executing the test plugin script.  There's no other way to debug into appc/cni than to re-vendor it to myself on github so I can add debugging prints so that Shippable spits out usable information, because the Travis tests verify that the Godeps are the same as upstream, and I cannot change appc/cni to check.

My current theory is that we're hitting https://github.com/golang/go/issues/9173 which is only fixed in Go 1.5, but it's been exceedingly hard to debug because (a) shippable takes forever to complete a test and (b) we can't get any intermediate output from Shippable and (c) we can't easily debug into appc/cni because Go really, really doesn't like changing godeps where the changes are not upstream on github.

thockin commented:
status?

dcbw commented:
@thockin I'm testing my theory with a modification to appc that works around the EPIPE thing.  If that passes Shippable I'll submit the PR for appc/cni and hopefully get @eyakubovich to merge it officially, and then I can ditch all my Godeps changes.  Shippable is being quite slow today...

dcbw commented:
I signed it!

thockin commented:
status?  Is it worth kicking the jenkins build again?

dcbw commented:
@thockin it seems to be working in Shippable now at least, I've backed out all the testing commits.  It seems we can work around the issue of the EPIPE in Kubernetes with some magic in the test script.

dcbw commented:
@thockin I've rebased on top of latest master and added another sanity check to the test script; hopefully this is the end of things.  I think the Jenkins build may have failed due to upstream changes that just got reverted, but I can't actually see Jenkins (I get "access denied").

thockin commented:
!!! 'gofmt -s' needs to be run on the following files: 
./pkg/kubelet/network/cni/cni_test.go

dcbw commented:
@thockin Travis and shippable are green now so I think we're good to go.  Suggestions about the CLA thing appreciated; both my and Rajat's addresses are @redhat so we are both covered on the CLA.

thockin commented:
@k8s-bot test this please

thockin commented:
I've already poked our CLA folks.

thockin commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1190,6 +1192,32 @@ func (dm *DockerManager) PortForward(pod *kubecontainer.Pod, port uint16, stream
 }
 
+// Get the IP address of a container's interface using nsenter
+func (dm *DockerManager) GetContainerIP(containerID, interfaceName string) (string, error) {
```
For future fixup: Could this just be private to the CNI plugin?

dcbw replied:
> For future fixup: Could this just be private to the CNI plugin?

Yeah, it probably could be.  I think Rajat just put in the docker stuff for symmetry, but I think in the future this code will go away when we get the IP from AddNetwork or add a STATUS call to the CNI spec.

dcbw commented:
@thockin Sorry this one took so long around the release.  I hope my future contributions will have a smoother landing.

dcbw commented:
I signed it!

thockin commented:
@k8s-bot test this please

thockin commented:
@k8s-oncall  The CLA bot is being a PITA - manually confirmed CLA, I'd like to get this merged

fabioy commented:
Have to improve the CLA thing. Set it to cla:yes and waiting for merge bot.

thockin commented:
@k8s-bot test this please

thockin commented:
Lawyers say CLA is OK on this one.

thockin commented:
I'm manually merging it.

jojimt commented:
Is there a released package that includes CNI? When I install with source_type == "packageManager", the kubelet does not seem to recognize CNI. If I build the binaries locally, however, it does recognize it.
